manim/config.py

```diff
@@ -1,326 +1,413 @@
-            dir_config["media_dir"] = "./media"
-        else:
-            print(
-                f"Media will be written to {dir_config['media_dir'] + os.sep}. You can change "
-                "this behavior with the --media_dir flag, or by adjusting dirs.py.,"
-            )
-    else:
-        if config["media_dir"]:
-            print(
-                "Ignoring --media_dir, since both --tex_dir and --video_dir were passed."
-            )
-
-    dir_config["tex_dir"] = (config["tex_dir"]
-                             or dirs.TEX_DIR
-                             or os.path.join(dir_config["media_dir"], "Tex"))
-    dir_config["text_dir"] = (config["text_dir"]
-                              or dirs.TEXT_DIR
-                              or os.path.join(dir_config["media_dir"], "texts"))
-
-    if not config["video_dir"] or dirs.VIDEO_DIR:
-        dir_config["video_dir"] = os.path.join(dir_config["media_dir"], "videos")
-
-    for folder in [dir_config["video_dir"], dir_config["tex_dir"], dir_config["text_dir"]]:
-        if folder != "" and not os.path.exists(folder):
+    # Specify the manim.cfg file
+    parser.add_argument(
+        "--config_file",
+        help="Specify the configuration file",
+    )
+
+    return parser.parse_args(arg_list)
+
+
+def _init_dirs(config):
+    # Make sure all folders exist
+    for folder in [config["media_dir"], config["video_dir"],
+                   config["tex_dir"], config["text_dir"]]:
+        if not os.path.exists(folder):
             os.makedirs(folder)
 
-    dirs.MEDIA_DIR = dir_config["media_dir"]
-    dirs.VIDEO_DIR = dir_config["video_dir"]
-    dirs.TEX_DIR = dir_config["tex_dir"]
-    dirs.TEXT_DIR = dir_config["text_dir"]
-
-def register_tex_template(tpl):
-    """Register the given LaTeX template for later use.
-
-    Parameters
-    ----------
-    tpl : :class:`~.TexTemplate`
-        The LaTeX template to register.
-    """
-    constants.TEX_TEMPLATE = tpl
-
-def initialize_tex(config):
-    """Safely create a LaTeX template object from a file.
-    If file is not readable, the default template file is used.
-
-    Parameters
-    ----------
-    filename : :class:`str`
-        The name of the file with the LaTeX template.
-    """
-    filename=""
-    if config["tex_template"]:
-        filename = os.path.expanduser(config["tex_template"])
-    if filename and not os.access(filename, os.R_OK):
-        # custom template not available, fallback to default
-        logger.warning(
-            f"Custom TeX template {filename} not found or not readable. "
-            "Falling back to the default template."
-        )
-        filename = ""
-    if filename:
-        # still having a filename -> use the file
-        constants.TEX_TEMPLATE = TexTemplateFromFile(filename=filename)
-    else:
-        # use the default template
-        constants.TEX_TEMPLATE = TexTemplate()
+
+def _from_command_line():
+    """Determine if manim was called from the command line."""
```

Comment:
Are we sure this wouldn't break with aliases? Can't we just `if __name__ == "__main__"`?

Reply:
It does break aliases. We can't just `if __name__ == "__main__"`.

Reply:
Why can't we just `if __name == "__main__"`? This is usually the way used in python to determine whether or not the program was invoked independently

Reply:
Yes thank you, I know that. The reason is that `__main__` will only be set to `__name__` inside the `__main__.py` module, and it will be something else inside the `config.py` module. And there is no easy way of signaling this all the way to `config.py`. 

Or at least I haven't figured out a way. I'm all ears if you have ideas :)

Reply:
We can probably let the `__main__.py` module signal this by a function or global variable or something like that. That might be the best way to go

Reply:
Could you please be more specific? I don't know how a function would help
here since `__main__.py` is importing `config.py`, and there is no way of
sending information to the imported module. If by global variable you mean
that `__main__.py` could export a variable that then `config.py` would
read: I'd hesitate to do that because in that case each module would be
importing the other one and I'm not sure that's necessary.

The current way is very hacky and ugly but it works and it doesn't involve
circular imports.

Reply:
@PgBiel responded:

> we could create a file in utils or something with a function called set_is_main or set_main, which sets the variable is_main that would be exported by this file

Reply:
Ok, this is interesting. I think this would make the config setup a bit slower (and it's already slow enough as it is). I'm on the fence about this as I don't like the current setup, but this seems awkward as well.

Honestly at this point, I'd want to ask over on SO since somebody must have had this problem before, right?

@eulertour your thoughts?

Reply:
Huh? I'm pretty sure this wouldn't make it any slower at all

Reply:
Ok, so the workflow is:
1. user call manim from command line, which executes `__init__.py`.
2. `__init__.py` imports `utils/is_main.py`, which sets the variable `IS_MAIN` to `True` and exports it.
3. `__init__.py` then import `config.py` and `config.py` imports `utils/is_main.py` in turn. `config.py` can now read `IS_MAIN` as being `True`.

Question: when/how does `IS_MAIN` ever get set to `False`?

Reply:
it is `False` by default

Reply:
on `is_main.py` we define a variable `IS_MAIN = False` and a function `set_is_main` which sets it to `True`. If `__main__` imports it, it will proceed to run `set_is_main`, which lets other files see that `IS_MAIN` is True. Otherwise, it will remain False as it always was, and it is assumed that the program was not called via the command line.

Reply:
I still don't understand. So `utils/is_main.py` looks like this:

```
__all__ = ['IS_MAIN']
IS_MAIN = False
if <something_here>:
    IS_MAIN = True
```

What goes in the if statement? It has to detect whether `utils/is_main.py` is being imported from `__main__.py`, but I don't know how to do this except in the hacky way that is currently implemented.

Am I missing something here?

Reply:
is_main.py:
```py
__all__ = ['IS_MAIN', 'set_is_main`]
IS_MAIN = False
def set_is_main():
    IS_MAIN = True
```

`__main__.py`:

```py
from .utils.is_main import set_is_main
set_is_main()
```

Reply:
HOLY SHIT THIS WORKS wtf python ok @PgBiel you're a genius

Reply:
lol thx
we should probably also guard that with `if __name__ == "__main__"` but that example is good enough to show what the main idea is

Reply:
So unfortunately this is not going to work. The reason is that `__main__.py` is actually not the first module that is executed. The first thing that is executed is always `__init__.py`. And the first line of code in `__init__.py` file is to import `config.py` (this cannot be changed).

The fix would be to have `__init__.py` set the `IS_MAIN` variable of `utils/is_main.py` correctly. HOWEVER, the `__name__` of `__init__.py` is NEVER equal to `"__main__"` and therefore there is no way to determine whether we need to call `set_is_main` from within `__init__.py`. Only `__main__.py` can do it, but by the time that `__main__.py` is called, `__init__.py` already set the config.

Reply:
I don't get it, how is `__init__` the first file to be executed? And why is the `__name__` not `__main__` then - what is it then? Note that `__name__ == '__main__'` does not require the file to be named `__main__`

Reply:
I was also stumped by this, but no matter how I run it, `__init__.py` is always the first thing to run. I think it's just how python packages are organized. The `__name__` variable within `__init__.py` is set to `"manim"`.

Reply:
That's probably only in the case of `python -m manim`. In that case, we can check if `__name__ in ["manim", "__main__"]` and be good to go

Reply:
1. It is not only the case when ran with `python -m`.
2. The `__name__` variable of `__init__.py` is `manim` every time it is imported, not only when it is being ran from the console as either `python -m manim` or `manim`. And the whole point of this is to have different behaviors when it is being imported vs when it is being ran from the console.

Also, I feel like I'm repeating myself and we are not seeing eye to eye on this one. You've been suggesting changes that don't actually work, and I've been spending my time implementing them. So if you come up with a solution _that actually works_, I would very much appreciate it if you implemented it yourself and posted it here. Thanks.